cmd/kaspawallet/daemon/server/split_transaction.go

```diff
@@ -0,0 +1,199 @@
+package server
+
+import (
+	"github.com/kaspanet/go-secp256k1"
+	"github.com/pkg/errors"
+
+	"github.com/kaspanet/kaspad/cmd/kaspawallet/libkaspawallet"
+	"github.com/kaspanet/kaspad/cmd/kaspawallet/libkaspawallet/serialization"
+	"github.com/kaspanet/kaspad/domain/consensus/model/externalapi"
+	"github.com/kaspanet/kaspad/domain/consensus/utils/consensushashing"
+	"github.com/kaspanet/kaspad/domain/consensus/utils/constants"
+	"github.com/kaspanet/kaspad/domain/consensus/utils/utxo"
+	"github.com/kaspanet/kaspad/domain/miningmanager/mempool"
+	"github.com/kaspanet/kaspad/util"
+)
+
+func (s *server) maybeAutoCompoundTransaction(transactionBytes []byte, toAddress util.Address,
+	changeAddress util.Address, changeWalletAddress *walletAddress) ([][]byte, error) {
+	transaction, err := serialization.DeserializePartiallySignedTransaction(transactionBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	splitTransactions, err := s.maybeSplitTransaction(transaction, changeAddress)
+	if err != nil {
+		return nil, err
+	}
+	if len(splitTransactions) > 1 {
+		mergeTransaction, err := s.mergeTransaction(splitTransactions, transaction, toAddress, changeAddress, changeWalletAddress)
+		if err != nil {
+			return nil, err
+		}
+		splitTransactions = append(splitTransactions, mergeTransaction)
+	}
+
+	splitTransactionsBytes := make([][]byte, len(splitTransactions))
+	for i, splitTransaction := range splitTransactions {
+		splitTransactionsBytes[i], err = serialization.SerializePartiallySignedTransaction(splitTransaction)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return splitTransactionsBytes, nil
+}
+
+func (s *server) mergeTransaction(
+	splitTransactions []*serialization.PartiallySignedTransaction,
+	originalTransaction *serialization.PartiallySignedTransaction,
+	toAddress util.Address,
+	changeAddress util.Address,
+	changeWalletAddress *walletAddress,
+) (*serialization.PartiallySignedTransaction, error) {
+	numOutputs := len(originalTransaction.Tx.Outputs)
+	if numOutputs > 2 || numOutputs == 0 {
+		return nil, errors.Errorf("original transaction has %d outputs, while 1 or 2 are expected",
+			len(originalTransaction.Tx.Outputs))
+	}
+
+	totalValue := uint64(0)
+	sentValue := originalTransaction.Tx.Outputs[0].Value
+	utxos := make([]*libkaspawallet.UTXO, len(splitTransactions))
+	for i, splitTransaction := range splitTransactions {
+		output := splitTransaction.Tx.Outputs[0]
```

Comment:
We don't appear to be merging the change. Is it not something we want to do at this stage?

Reply:
The sum of the outputs of the splits is equal to the total of their inputs. This includes the change as well.

Change is re-calculated again in mergeTransaction:
```
	if totalValue > sentValue {
		payments = append(payments, &libkaspawallet.Payment{
			Address: changeAddress,
			Amount:  totalValue - sentValue,
		})
``` 
(Note the change is going to be a bit lower, because there's fee for the mergeTransaction)